Модуль 7

AntonovIgor opened the pull request.

AntonovIgor committed:
7.1. Добавит `LocalStrategy` для Passport. WIP

Потренируемся работать со стратегиями аутентификации Passport. Проверка
логина и пароля происходит прямо в обработчике маршрута `/auth/login`.
Мы используем Passport, поэтому логично её реализовать в виде отдельной
стратегии — `LocalStrategy`. Ранее мы создавали стратегию для проверки
access token (далее AT), теперь реализуем похожую, но только для проверки
логина и пароля.

Создадим отдельный модуль `local.strategy.ts` и в нём опишем новый класс
`LocalStrategy`. Не забудьте воспользоваться наследованием от
`PassportStrategy`.

Этому классу потребуется доступ к сервису аутентификации (`AuthService`),
поэтому сразу внедрим его в качестве зависимости. Внутри конструктора не
забудем вызвать родительский конструктор. Аргументом ему нужно передать
объект настроек. На текущем этапе из настроек нас интересует только
переопределение `usernameField`.

В этом поле необходимо передать информацию о названии поля из которого
извлекается логин. В нашем случае это `email`.

Последним шагом сделаем реализацию метода `validate`. Она предельно
проста, так внутри требуется вызвать лишь метод `verifyUser` сервиса
`AuthenticationService`.

WIP: Интеграция `LocalStrategy` не завершена.

AntonovIgor committed:
7.2. Подключит стратегию `LocalStrategy` в `AuthenticationModule`. WIP

Чтобы воспользоваться новой стратегией в контроллере
`AuthenticationController` добавим её в качестве провайдера в модуль
`AuthenticationModule`.

WIP: Интеграция `LocalStrategy` продолжается.

AntonovIgor committed:
7.3. Добавит `LocalAuthGuard`. WIP

Чтобы воспользоваться `LocalStrategy` нам потребуется новый `Guard`.
Опишем его в модуле `local-auth.guard.ts`. Имплементация похожа на guard,
для JWT.

`AuthGuard` примет запрос и передаст данные для аутентификации в
Passport.

WIP: Интеграция `LocalStrategy` продолжается.

AntonovIgor committed:
7.4. Подключит `LocalAuthGuard` в `AuthenticationController`

Для ресурса `/auth/login` добавим поддержку `LocalAuthGuard`.Воспользуемся
декоратором `@UseGuards` и аргументом передадим ему созданные в прошлом шаге
`LocalAuthGuard`.

Внутри обработчика извлечём информацию о пользователе из запроса. Для этого
воспользуемся декоратором `@Req`. Последним шагом вернём информацию о
токенах авторизованного пользователя. Попробуйте запустить приложение
на этом шаге и протестировать обработчик маршрута `/login`.

Проверка логина и пароля выполняется через локальную стратегию авторизации.

AntonovIgor committed:
7.5. Добавит настройки для формирования refresh token

Для формирования Refresh Token используется отдельный секрет. Добавим
в конфигурацию дополнительные настройки. Через них передадим секрет и
время жизни Refresh Token.

Важно: не забудьте задать значения для новых настроек в `account.env`

AntonovIgor committed:
7.6. Добавит формирование Refresh Token.

В `AuthenticationService` добавим подготовку Refresh Token. Токен для
обновления формируется точно таким же образом, как и AccessToken. С
одним лишь исключением: для формирования используются другие настройки
(отдельный секрет и время жизни).

AntonovIgor committed:
7.7. Добавит `JwtRefreshStrategy` и `JwtRefreshGuard`

Добавим ещё одну стратегию для Passport. На этот раз в ней будем работать
с Refresh Token. Общий принцип такой же, как и при создании стратегии
JWT. RefreshToken извлечём из заголовка `Authorization`. В методе
`validate` будет доступен `payload` токена. На основе `payload` получим
пользователя из базы данных.

Также по аналогии подготовим `AuthGuard`. Только для стратегии `jwt-refresh`.

AntonovIgor committed:
7.8. Добавит `refreshToken` в `AuthenticationController`

Последним шагом обновим обработчик `refresh` в `AuthController`. Попробуйте
запустить приложение, выполнить вход, а затем получить новую пару токенов на
основании refresh-токен.

Приложение работает, но есть проблема. Время жизни RefreshToken большое и
после получения новой пары токенов, старый RefreshToken продолжить жить.
Следовательно, на его основании можно получить ещё одну пару токенов.

Это проблема и чтобы её решить, потребуется предпринять дополнительные действия.

ratelChief commented:
@AntonovIgor  а когда появятся ченджи, показывающие примеры реализации для блога/юзера?

AntonovIgor committed:
7.9. Добавит интерфейс `JwtToken`.

Базовая поддержка двух видов токенов реализована. Добавим отдельный
интерфейс `JwtToken`. Он потребуется в дальнейшем для определения модели
и сущности (Entity).

А зачем нам может потребоваться модель и сущность для токенов? Они
же формируются динамически. Напомним, Refresh Token реализованы,
но сейчас по одному Refresh Token можно получить сколько угодно
пар из Access и Refresh Token. Помимо этого, нет в случае
утечки RT, злоумышленник может сгенерировать целую пачку токенов
и есть только один способ разрешить эту ситуацию: изменить секрет,
который применяется для формирования токенов.

Способ рабочий, но тогда у всех пользователей приложения слетит
авторизация. Все токены станут невалидными.

Более элегантный способ заключается в хранении выданных Refresh Token
в базе данных. Когда клиент приходит за новой парой токенов,
выполняется проверка наличия Refresh Token в базе. Если он
есть, выписывается новая пара токенов, а старый Refresh Token
удаляется из базы. На его место приходит новый RT.

Реализовав выше описанный алгоритм, дважды получить новую
пару токенов на основании одного Refresh Token не получится.

Разберём решение этой задачи на практике. Определим новый
интерфейс.

AntonovIgor committed:
7.10. Добавит тип `RefreshTokenPayload`

В полезной нагрузке Refresh Token потребуется дополнительная
информация — идентификатор токена. По нему будем производить поиск
токена в базе данных. Создадим новый интерфейс `RefreshTokenPayload`
и унаследуем его от существующего `TokenPayload`.

AntonovIgor committed:
7.11. Добавит функцию `parseTime`

Добавим вспомогательную функцию `parseTime`. С её помощью разберём строку, содержащую время жизни токена.

Также добавим вспомогательную функцию `createJWTPayload`. Она упростит подготовку полезной нагрузки токена.

AntonovIgor committed:
7.12. Добавит модуль управления Refresh Token

Для хранения информации о Refresh Token в базе подготовим все необходимые элементы: entity, model, repository…

AntonovIgor committed:
7.13. Добавит определение ошибки `TokenNotExistsException`

Nest не ограничивает нас использованием только предопределённых ошибок. Мы можем создавать собственные ошибки, наследуя от существующих.

AntonovIgor committed:
7.14. Обновит стратегию `JwtRefreshStrategy`

Помимо валидации токена и проверки его времени жизни, добавим проверку на наличие токена в базе данных. Если токена нет, то бросим исключение `TokenNotExistsException`.

Если проверка пройдена, удалим старый Refresh Token, а также удалим все возможные просроченные токены.

AntonovIgor committed:
7.15. Обновит метод `createUserToken`

Обновим в `authentication.service.ts` метод `createUserToken`. Добавим поддержку сохранения RT в базе данных и формирование `tokenId`.

ratelChief commented:
@AntonovIgor заметил, что в юзерах, наш метод из репозитория не возвращает id, а возвращает _id из монги, что делает невозможным работу login метода, советую проверить
